src/ossl/aes.rs

```diff
@@ -911,84 +988,491 @@ impl AesOperation {
+
+        /* reset ctx */
+        let res = unsafe { EVP_CIPHER_CTX_reset(self.ctx.as_mut_ptr()) };
+        if res != 1 {
+            return Err(CKR_DEVICE_ERROR)?;
+        }
+
+        self.encrypt_initialize()?;
+
+        if self.params.iv.gen != CKG_NO_GENERATE {
+            let iv = bytes_to_slice!(mut iv_ptr, self.params.iv.buf.len(), u8);
+            iv.copy_from_slice(&self.params.iv.buf);
+        }
+
+        #[cfg(feature = "fips")]
+        self.fips_approval_aead()?;
+
+        Ok(())
+    }
+
+    pub fn msg_decrypt_init(
+        mech: &CK_MECHANISM,
+        key: &Object,
+    ) -> Result<AesOperation> {
+        Ok(AesOperation {
+            mech: mech.mechanism,
+            op: CKF_MESSAGE_DECRYPT,
+            key: object_to_raw_key(key)?,
+            /* params are not set until later */
+            params: Self::init_params(&CK_MECHANISM {
+                mechanism: CK_UNAVAILABLE_INFORMATION,
+                pParameter: std::ptr::null_mut(),
+                ulParameterLen: 0,
+            })?,
+            finalized: false,
+            in_use: false,
+            ctx: EvpCipherCtx::new()?,
+            finalbuf: Vec::new(),
+            blockctr: 0,
+            #[cfg(feature = "fips")]
+            fips_approved: None,
+        })
+    }
+
+    fn msg_decrypt_new(
+        &mut self,
+        parameter: CK_VOID_PTR,
+        parameter_len: CK_ULONG,
+        aad: &[u8],
+    ) -> Result<()> {
+        if self.op != CKF_MESSAGE_DECRYPT {
+            return Err(CKR_OPERATION_NOT_INITIALIZED)?;
+        }
+        if self.in_use {
+            return Err(CKR_OPERATION_ACTIVE)?;
+        }
+
+        #[cfg(feature = "fips")]
+        {
+            self.params.zeroize();
+            self.finalbuf.zeroize();
+        }
+        self.finalized = false;
+        self.in_use = true;
+
+        let _ = self.init_msg_params(parameter, parameter_len, aad)?;
+
+        /* reset ctx */
+        let res = unsafe { EVP_CIPHER_CTX_reset(self.ctx.as_mut_ptr()) };
+        if res != 1 {
+            return Err(CKR_DEVICE_ERROR)?;
+        }
+
+        self.decrypt_initialize()?;
+
+        #[cfg(feature = "fips")]
+        self.fips_approval_aead()?;
+
+        Ok(())
+    }
+
+    #[cfg(feature = "fips")]
+    fn fips_approval_aead(&mut self) -> Result<()> {
+        /* For AEAD we handle indicators directly because OpenSSL has an
+         * inflexible API that provides incorrect answers when we
+         * generate the IV outside of that code */
+
+        /* The IV size must be 12 in FIPS mode */
+        if self.params.iv.buf.len() != 12 {
+            self.fips_approved = Some(false);
+            return Ok(());
+        }
+
+        /* The IV must be generated in FIPS mode */
+        self.fips_approved = match self.params.iv.gen {
+            CKG_NO_GENERATE => match self.op {
+                CKF_MESSAGE_ENCRYPT => Some(false),
+                CKF_MESSAGE_DECRYPT => Some(true),
```

Comment:
I think that this is not how it works. I think the requirement about generation is about the advancing of the IV from the initial value and generating the following IVs for next blocks, not about the providing the initial IV altogether.

Reply:
Nope, I followed how softoken classifies things, and really you have to generate the IV on the token. OpenSSL does the same for TLS with a special layer breaking cipher mode that does not allow to pass in an IV at all (if you do pass an IV, in older implementations, it returns an error and terminates the operation).

Reply:
So this is for the TLS. In SSH, we have KDF, which generates the GCM IV and it needs to set it somehow to the cipher context, otherwise it would not be FIPS compliant. What is the way how we will make it FIPS compliant for SSH?

Reply:
Some protocols have protocol-specific exceptions. In fact TLS does *not* generate the IV inside the module, which is why thee is an exception and a specific interface in OpenSSL for it (sorry I realize I was not clear in the comment above).

So for general use of GCM, FIPS-140-3 mandates very strict rules, and IV generation inside the token is one of them, for TLS specifically there is an exception where part of the IV can be provided to the module. 

I think the implementation I have in this patch-set is FIPS compliant, (it's the whole point of even adding this rarely used interface) but if FIPS certification labs will find something to object we will add further patches later.

Reply:
Ok, thanks for the clarification.